src/krux/touch.py

```diff
@@ -0,0 +1,111 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import time
+from .touchscreens.ft6x36 import FT6X36
+
+TOUCH_R_PERIOD = 50  # 1000/50 = 20 samples/second
+
+
+class Touch:
+    """Touch is a singleton API to interact with touchscreen driver"""
+
+    idle, press, release = 0, 1, 2
+
+    def __init__(self, width, height):
+        """Touch API init - width and height are in Landscape mode
+        For Krux width = max_y, height = max_x
+        """
+        self.cycle = TOUCH_R_PERIOD
+        self.last_time = 0
+        self.y_regions = []
+        self.x_regions = []
+        self.index = 0
+        self.state = Touch.idle
+        self.width, self.height = width, height
+        self.touch_driver = FT6X36()
+
+    def clear_regions(self):
+        """Remove previously stored buttons map"""
+        self.y_regions = []
+        self.x_regions = []
+
+    def add_y_delimiter(self, region):
+        """Adds a y button delimiter to be mapped as a array by touchscreen"""
+        if region > self.width:
+            raise ValueError("Touch region added outside display area")
+        self.y_regions.append(region)
+
+    def add_x_delimiter(self, region):
+        """Adds a x button delimiter to be mapped as a array by touchscreen"""
+        if region > self.height:
+            raise ValueError("Touch region added outside display area")
+        self.x_regions.append(region)
+
+    def extract_index(self, data):
+        """Gets an index from touched points, y and x delimiters"""
+        if self.state != Touch.press:
+            self.state = Touch.press
+            self.index = 0
+            if self.y_regions:
+                for region in self.y_regions:
+                    if data[1] > region:
+                        self.index += 1
+                if self.index == 0 or self.index >= len(
+                    self.y_regions
+                ):  # outside y areas
+                    self.state = Touch.idle
+                else:
+                    self.index -= 1
+                    if self.x_regions:  # if 2D array
+                        self.index *= len(self.x_regions) - 1
+                        x_index = 0
+                        for x_region in self.x_regions:
+                            if data[0] > x_region:
+                                x_index += 1
+                        if x_index == 0 or x_index >= len(
+                            self.x_regions
+                        ):  # outside x areas
+                            self.state = Touch.idle
+                        else:
+                            x_index -= 1
+                        self.index += x_index
+            else:
+                self.index = 0
+
+    def current_state(self):
+        """Returns the touchscreen state"""
+        if time.ticks_ms() > self.last_time + self.cycle:
```

Comment:
Testing the build locally on my Amigo now and I noticed that you have to long-press to get button presses to work. I made a new build where I removed this timer condition (so `current_state` is now always checking when called) and it made touches more responsive. Could you make that change here? Both `self.cycle`  and `self.last_time` can be removed

Reply:
Before I do this: Did you let it running like this for a while? On my early tests with touch it would crash(IO error) if I didn't limit this I2C pulling frequency, maybe there was another cofactor on this issue that was solved later, but I think it is a good idea to let it running for some minutes, keep navigating on menus, run some QR scans, etc. to check if it remains stable.

Reply:
I tested here, and after 5 min running it happened:
```
Traceback (most recent call last):                                                                                                
  File "boot.py", line 42, in <module>                                                                                            
  File "krux/pages/__init__.py", line 315, in run                                                                                 
  File "krux/pages/__init__.py", line 334, in run_loop                                                                            
  File "krux/input.py", line 80, in wait_for_button                                                                               
  File "krux/input.py", line 76, in wait_for_press                                                                                
  File "krux/input.py", line 58, in touch_value                                                                                   
  File "krux/touch.py", line 70, in value                                                                                         
  File "krux/touch.py", line 60, in current_state                                                                                 
  File "krux/touchscreens/ft6x36.py", line 30, in current_point                                                                   
  File "krux/touchscreens/ft6x36.py", line 26, in read_reg                                                                        
OSError: [Errno 5] EIO                                                                                                            
MicroPython v1.11 on 2022-04-21; Sipeed_M1 with kendryte-k210                                                                     
Type "help()" for more information.        
```
Maybe there are ways to prevent this, like checking for a busy I2C, but I don't know exactly what is causing this, for now I would let it the way it is, maybe the sample rate can be increased a bit, but I would test well before doing it.

Reply:
Strange, I actually used it for a while (much longer than 5 minutes) and didn't run into this error. Nonetheless, if it's an intermittent error, maybe we can just keep track of the last point successfully read from the register and wrap reads in `current_point` in a try..catch:

```
    def current_point(self):
        """If touch is pressed, returns x and y points"""
        try:
            data = self.read_reg(0x02, 1)
            if data is not None and data[0] == 0x1:
                data_buf = self.read_reg(0x03, 4)
                x = ((data_buf[0] & 0x0F) << 8) | (data_buf[1])
                y = ((data_buf[2] & 0x0F) << 8) | (data_buf[3])
                if (data_buf[0] & 0xC0) == 0x80:
                    self.last_point = (x, y)
                    return self.last_point
        except:
            return self.last_point

        self.last_point = None
        return self.last_point
```
and in  `__init__`:

`self.last_point = None`

Can you give that a try and see if it fixes your problem?